<center><h3><font color='#E95420'>Optimization Methods for Machine Learning</font></h3>
<center><h3><font color='#772953'>Homework 1</font></h3></center>
<br><br> 
<center><h4><font color='#333333'>Michele Cernigliaro</font></h4></center>
<center><h4><font color='#333333'>Francesco Russo</font></h4></center>
<br><br>

<center><h4><b><font color='#333333'>Question 1</font></b></h4></center>

You must construct a shallow **Feedforward Neural Network** (FNN) (one only hidden layer), either
a **MLP** or a **RBF** network, that provides the model $f(x)$. We denote by ***π*** the hyper-parameters of
the network to be settled by means of an heuristic procedure and ***ω*** the parameters to be settled
by minimizing the regularized training error.

<center><h4><b><font color='#333333'>Question 1.1)</font></b></h4>
___________________________________________________________________________________________________
</center>
<br><br>

We must construct a **MLP** (Multi Layer Perceptron)

In [1]:
import pandas as pd
import numpy as np

<center>
$\displaystyle{ f(x) = \sum_{j = 1}^N \left[v_j g \left(\sum_{i = 1}^2 w_{ji} \cdot x_i - b_j \right) \right] }$
<br><br>
\[\vec{x_1} = \begin{bmatrix}
x_{11} \\ ... \\ x_{1P}
\end{bmatrix}
\quad\quad
\vec{x_2} = \begin{bmatrix}
x_{21} \\ ... \\ x_{2P}
\end{bmatrix}
\quad\quad
\vec{y} = \begin{bmatrix}
y_{1} \\ ... \\ y_{P}
\end{bmatrix}
\quad\quad
\vec{\pi} = \begin{bmatrix}
N \\ \sigma \\ \rho
\end{bmatrix}
\quad\quad
\vec{\omega} = \begin{bmatrix}
v \\ w \\ b
\end{bmatrix}
\]
<br>
$\displaystyle{ g(t) = tanh(t) = \frac{e^{2 \sigma t} - 1}{e^{2 \sigma t} + 1} }$
<br><br>
$\displaystyle{ E(\omega; \pi) = \frac{1}{2P} \sum_{p = 1}^P \left[ f(x_p) - y_p \right]^2 + \rho \cdot ||\omega||^2 }$
<img src="FFNN.png">
</center>

In [2]:
file = pd.ExcelFile('dataPoints.xlsx')
df = file.parse('Foglio1')
df.head()

x1        x2         y
0 -1.262271  0.746859  0.461002
1 -0.804355  0.937081  0.617336
2 -1.805674  0.738389 -0.053984
3 -0.888805  0.061711  1.943678
4 -1.667562 -0.534543  2.127070

In [3]:
X = df[['x1', 'x2']].to_numpy()
X = np.append(X, -1 * np.ones((X.shape[0], 1)), axis=1)
y_true = df[['y']].to_numpy().reshape(1,-1)

In [4]:
def g_fun(X):
    return np.tanh(X)

def fun(X, omega):
    # extract param vectors from omega
    v = omega[0:N].reshape(1, N)
    W = omega[N:3*N].reshape(N, n)
    b = omega[3*N:].reshape(N, 1)
    
    # merge W with b
    Wb = np.append(W, b, axis=1)
    
    return np.dot(v, g_fun(np.dot(Wb, np.transpose(X))))

In [5]:
#np.random.seed(123) # seed (for reprudecebility)

P = X.shape[0]        # number of observations
N = 10                # number of neurons in hidden layer
n = X.shape[1] - 1    # number of independent variables
sigma = 1
rho = 0.0005

# parameters initialization
W = np.random.random((N,n))
b = np.random.random((N,1))
v = np.random.random((N,1))

#print(v, "\n")
#print(W, "\n")
#print(b, "\n", '-------------------')

# all the parameters
omega = np.concatenate((v, W.reshape(N*n,1), b))

#print(omega[0:N].reshape(1, N),'\n')
#print(omega[N:3*N].reshape(N, X.shape[1]),'\n')
#print(omega[3*N:].reshape(N, 1))

In [6]:
def loss(omega):
    y_pred = fun(X, omega)
    l = np.sum((y_pred - y_true)**2)/(2 * 300) + rho * np.linalg.norm(omega)**2
    return l

In [7]:
loss(omega)

6.037605870646368

In [8]:
from scipy.optimize import minimize as minimize

In [9]:
q = minimize(loss, omega, method = "BFGS", options = {'gtol': 1e-05, 'eps': 1.4901161193847656e-08})
q

      fun: 0.06976955622313348
 hess_inv: array([[ 1.26964095e+02,  6.36587515e+01, -5.04152812e+01, ...,
        -1.09001980e+01,  1.30584003e+01,  2.49970447e+00],
       [ 6.36587515e+01,  8.99008135e+02,  3.45211410e+00, ...,
        -8.42952288e-01, -4.65988472e+01, -1.53845094e+01],
       [-5.04152812e+01,  3.45211410e+00,  1.63048978e+02, ...,
         8.46767078e+00,  2.27156615e+01,  2.68639027e+01],
       ...,
       [-1.09001980e+01, -8.42952288e-01,  8.46767078e+00, ...,
         2.47791349e+01,  2.56212011e+00, -1.30132341e+00],
       [ 1.30584003e+01, -4.65988472e+01,  2.27156615e+01, ...,
         2.56212011e+00,  1.58606542e+02,  9.71342023e+00],
       [ 2.49970447e+00, -1.53845094e+01,  2.68639027e+01, ...,
        -1.30132341e+00,  9.71342023e+00,  5.08782193e+01]])
      jac: array([ 7.23730773e-06, -6.09736890e-06,  4.51039523e-06, -8.59610736e-07,
       -2.99792737e-06,  7.73090869e-06,  6.17746264e-06,  1.12969428e-06,
        4.05125320e-06,  5.10457903e-06,

In [10]:
q.x

array([ 2.14067457,  1.34901845,  2.25660943, -1.45429605,  1.3512087 ,
       -1.80460788, -1.34863418, -1.9373903 , -1.35083843,  1.76047654,
       -0.21633047, -2.60624335, -0.04565675,  0.25101517,  0.49006776,
        1.14383338,  1.96865574, -0.16504924, -0.04611526,  0.25265046,
        0.36306062, -1.13255412,  0.04593858, -0.25205513,  0.05271357,
        2.92733609,  0.04595268, -0.25196033,  2.04117977,  0.0158146 ,
        2.52533403, -0.88936964,  1.23279955, -0.86421354, -0.88750924,
        0.6663127 ,  0.88858967,  1.02934309,  0.88873084,  0.99569439])

In [11]:
v_0 = omega[0:N].reshape(1, N)
W_0 = omega[N:3*N].reshape(N, n)
b_0 = omega[3*N:].reshape(N, 1)

In [12]:
print(v_0, "\n")
print(W_0, "\n")
print(b_0, "\n")

[[0.65518427 0.87931567 0.90577331 0.50491343 0.00847125 0.97413178
  0.10466271 0.04304338 0.00190801 0.228636  ]] 

[[0.02703791 0.31997048]
 [0.3479972  0.21505721]
 [0.72984043 0.57219695]
 [0.1787696  0.53277701]
 [0.74974063 0.40204775]
 [0.33905773 0.16086841]
 [0.97250422 0.53033218]
 [0.71953231 0.70530451]
 [0.79810654 0.37480766]
 [0.91780235 0.27194346]] 

[[0.5146439 ]
 [0.75315707]
 [0.30379646]
 [0.48756334]
 [0.42155458]
 [0.73837601]
 [0.98611104]
 [0.78959806]
 [0.75294506]
 [0.3685829 ]] 



In [13]:
v = q.x[0:N].reshape(1, N)
W = q.x[N:3*N].reshape(N, n)
b = q.x[3*N:].reshape(N, 1)

print(v, "\n")
print(W, "\n")
print(b, "\n")

[[ 2.14067457  1.34901845  2.25660943 -1.45429605  1.3512087  -1.80460788
  -1.34863418 -1.9373903  -1.35083843  1.76047654]] 

[[-0.21633047 -2.60624335]
 [-0.04565675  0.25101517]
 [ 0.49006776  1.14383338]
 [ 1.96865574 -0.16504924]
 [-0.04611526  0.25265046]
 [ 0.36306062 -1.13255412]
 [ 0.04593858 -0.25205513]
 [ 0.05271357  2.92733609]
 [ 0.04595268 -0.25196033]
 [ 2.04117977  0.0158146 ]] 

[[ 2.52533403]
 [-0.88936964]
 [ 1.23279955]
 [-0.86421354]
 [-0.88750924]
 [ 0.6663127 ]
 [ 0.88858967]
 [ 1.02934309]
 [ 0.88873084]
 [ 0.99569439]] 



In [14]:
loss(omega)

6.037605870646368

In [15]:
loss(q.x)

0.06976955622313348

In [ ]:
import plotly.graph_objects as go
import numpy as np

x = np.random.uniform(-2, 2, 100000).reshape(-1,1)
y = np.random.uniform(-1, 1, 100000).reshape(-1,1)
X_points = np.append(x, -1*np.ones((x.shape[0],1)), axis=1)
z = fun(np.append(X_points, y, axis=1), q.x)

x = x.reshape(-1)
y = y.reshape(-1)
z = z.reshape(-1)

fig = go.Figure(data=[go.Scatter3d(
    x=x,
    y=y,
    z=z,
    mode='markers',
    marker=dict(
        size=2,
        color=z,                # set color to an array/list of desired values
        colorscale='Viridis',   # choose a colorscale
        opacity=0.8
    )
)])

# tight layout
fig.update_layout(margin=dict(l=0, r=0, b=0, t=0))
fig.show()

<center><h4><b><font color='#333333'>Computation of the gradient</font></b></h4></center>

<center>
$\displaystyle{ f(x) = \sum_{j = 1}^N \left[v_j g \left(\sum_{i = 1}^2 w_{ji} \cdot x_i - b_j \right) \right] }$
    <br>
$\displaystyle{ E(\omega; \pi) = \frac{1}{2P} \sum_{p = 1}^P \left[ f(x_p) - y_p \right]^2 + \rho \cdot ||\omega||^2 } = $
<br>
$ = \displaystyle{\frac{1}{2P} \sum_{p = 1}^P \left[ \sum_{j = 1}^N \left[v_j g \left(\sum_{i = 1}^2 w_{ji} \cdot x_{pi} - b_j \right) \right] - y_p \right]^2 + \rho \cdot \left( \sum_{j = 1}^N v_j^2 + \sum_{i = 1}^n \sum_{j = 1}^N w_{ij}^2 + \sum_{j = 1}^N b_j^2 \right)}$
    <br>
$\displaystyle{ g(t) = tanh(t) = \frac{e^{2 \sigma t} - 1}{e^{2 \sigma t} + 1} }$
</center>

<center>
$\vec{\nabla E} = \displaystyle{ \left [\frac{\partial{E}}{\partial{v_1}} \space, ... , \space \frac{\partial{E}}{\partial{v_N}} \space , \space \frac{\partial{E}}{\partial{w_{11}}} \space , ... , \space \frac{\partial{E}}{\partial{w_{nN}}} \space , \space \frac{\partial{E}}{\partial{b_1}} \space , ... , \space \frac{\partial{E}}{\partial{b_N}} \right]^T }$
<br>
<br>
$\displaystyle{ 
\frac{\partial{E}}{\partial{\space v_{j'}}} = \frac{1}{P} \sum_{p = 1}^P \left[ f(x_p) - y_p \right] \cdot g\left(\sum_{i=1}^n w_{j'i}x_{pi} - b_{j'} \right) + 2 v_{j'}\rho
}$
<br>
<br>
$\displaystyle{ 
\frac{\partial{E}}{\partial{\space w_{j'i'}}} = \frac{1}{P} \sum_{p = 1}^P \left[ f(x_p) - y_p \right] \cdot \left[ v_{j'} \frac{4 \sigma e^{2\sigma t}}{(e^{2\sigma t} + 1)^2} x_{pi'} \right] + 2 w_{j'i'}\rho
}$
<br>
<br>
$\displaystyle{ 
\frac{\partial{E}}{\partial{\space b_{j'}}} = \frac{1}{P} \sum_{p = 1}^P \left[ f(x_p) - y_p \right] \cdot \left[ - \space v_{j'} \frac{4 \sigma e^{2\sigma t}}{(e^{2\sigma t} + 1)^2}\right] + 2 b_{j'}\rho
}$
</center>

In [16]:
def fun_grad(omega):
    
    # pick parameters from omega one
    v = omega[0:N].reshape(N,1)
    W = omega[N:3*N].reshape(N, n)
    b = omega[3*N:].reshape(N, 1)
    
    
    Wb = np.append(W,b, axis=1)
    
    ### compute partial derivatives in a "pythonic" way:
    
    # for the sake of clearness, let's define some variables
    # (MATRIX) dot product between Wb and X
    T = np.dot(Wb, np.transpose(X)) # NxP matrix
    
    # derivative of g()
    g_der = 4 * sigma * np.exp(2 * sigma * T) / (np.exp(2 * sigma * T) + 1)**2

    dE_dv = 1 / P * np.dot(g_fun(T), np.transpose(fun(X, omega) - y_true)) + 2 * rho * v
    dE_db = 1 / P * np.dot(-1 * v * g_der, np.transpose(fun(X, omega) - y_true)) + 2 * rho * b
    
    # dealing with dE_dW
    mat1 = v * g_der * X[:,0]
    mat1 = np.dot(mat1, np.transpose(fun(X, omega) - y_true))
    mat2 = v * g_der * X[:,1]
    mat2 = np.dot(mat2, np.transpose(fun(X, omega) - y_true))

    fusion = np.append(mat1, mat2, axis=1)
    dE_dW = 1 / P * fusion + 2 * rho * W

    omega_gradient = np.concatenate((dE_dv, dE_dW.reshape(N*n,1), dE_db))
    return omega_gradient.reshape(-1)

In [17]:
fun_grad(omega).shape

(40,)

In [18]:
fun_grad(omega)

array([ 1.50829709e+00,  2.07267255e+00,  1.36452049e+00,  1.56746125e+00,
        1.51474878e+00,  2.04923796e+00,  2.04396512e+00,  1.98556974e+00,
        1.90290559e+00,  1.41020441e+00,  6.87733687e-01,  1.79254114e-02,
       -1.12536568e-01,  3.45160871e-02,  3.37138783e-02,  5.27285135e-02,
        3.08323286e-01, -3.98051814e-02, -5.49370253e-04,  1.11461660e-03,
       -9.99508130e-02,  6.43604358e-02, -6.08447622e-02,  3.91769424e-03,
       -1.79493437e-02, -5.88411557e-04, -1.52560064e-04,  4.81378740e-04,
       -2.79002801e-02,  2.69918642e-02,  1.46327010e+00,  1.27618158e+00,
        1.30549856e+00,  1.02703400e+00,  1.18883280e-02,  1.45227027e+00,
        9.39072020e-02,  4.80987486e-02,  2.86982643e-03,  2.71165702e-01])

In [19]:
omega.reshape(-1).shape

(40,)

In [21]:
from scipy.optimize import minimize
q = minimize(loss, omega, jac = fun_grad, method = "BFGS", options = {'gtol': 1e-05, 'eps': 1.4901161193847656e-08})
q

      fun: 0.06976959406183966
 hess_inv: array([[ 1.11059602e+02,  2.69568564e+00, -9.56951336e+00, ...,
        -1.26020622e+01,  1.77817786e+01,  4.92331524e-02],
       [ 2.69568564e+00,  5.22123144e+02,  4.94304543e+01, ...,
         1.63594785e+00, -3.33553615e+01, -4.52200530e+00],
       [-9.56951336e+00,  4.94304543e+01,  1.60449337e+02, ...,
        -1.57649120e-01,  3.86684093e+00,  1.59697233e+01],
       ...,
       [-1.26020622e+01,  1.63594785e+00, -1.57649120e-01, ...,
         2.45523415e+01, -7.95955787e-01,  2.98065731e-01],
       [ 1.77817786e+01, -3.33553615e+01,  3.86684093e+00, ...,
        -7.95955787e-01,  1.19994472e+02, -4.95915698e+00],
       [ 4.92331524e-02, -4.52200530e+00,  1.59697233e+01, ...,
         2.98065731e-01, -4.95915698e+00,  3.11810677e+01]])
      jac: array([-4.50188530e-06, -1.05881796e-06, -2.00819106e-06,  3.02300982e-06,
        3.44672491e-06, -3.81130329e-06,  2.60909620e-06, -2.17078023e-06,
        7.74420894e-07,  3.23188138e-06,

In [22]:
loss(omega)

6.037605870646368

In [23]:
loss(q.x)

0.06976959406183966

<center><h4><b><font color='#333333'>Question 1.2)</font></b></h4>
___________________________________________________________________________________________________
</center>
<br><br>

We must now construct a **RBF** (Radial Basis Function) neural network

<center>
$\displaystyle{ f(x) = \sum_{j = 1}^N \left[v_j \phi \left(||\vec{x} - \vec{c}_j ||\right) \right] }$
<br><br>
\[\vec{x_1} = \begin{bmatrix}
x_{11} \\ ... \\ x_{1P}
\end{bmatrix}
\quad\quad
\vec{x_2} = \begin{bmatrix}
x_{21} \\ ... \\ x_{2P}
\end{bmatrix}
\quad\quad
\vec{y} = \begin{bmatrix}
y_{1} \\ ... \\ y_{P}
\end{bmatrix}
\quad\quad
\vec{\pi} = \begin{bmatrix}
N \\ \sigma \\ \rho
\end{bmatrix}
\quad\quad
\vec{\omega} = \begin{bmatrix}
v \\ c
\end{bmatrix}
\]
<br>
$\displaystyle{ \phi(||\vec{x} - \vec{c}||) = e^{\displaystyle{\frac{||\vec{x} - \vec{c}||^2}{\sigma^2}}}}$
<br><br>
$\displaystyle{ E(\omega; \pi) = \frac{1}{2P} \sum_{p = 1}^P \left[ f(x_p) - y_p \right]^2 + \rho \cdot ||\omega||^2 }$
<img src="RBF.png">
</center>